File: web/src/api/gen-template.ts
import axios from 'axios';
import { GenTemplate, CreateGenTemplateRequest, UpdateGenTemplateRequest, GenHistory, GenerateFromTemplateRequest, GenerateFromTemplateResponse, GenModel } from '@/types/gen-template';

export const getGenTemplates = async (workspaceId: string, pageNum: number, pageSize: number, query: string) => {
  const response = await axios.get<GenTemplate[]>(
    `/api/v1/workspaces/${workspaceId}/gen-templates?pageSize=${pageSize}&pageNumber=${pageNum}&query=${query}`,
    { withCredentials: true }
  );
  return response.data;
};

export const getGenTemplate = async (workspaceId: string, templateId: string) => {
  const response = await axios.get<GenTemplate>(
    `/api/v1/workspaces/${workspaceId}/gen-templates/${templateId}`,
    { withCredentials: true }
  );
  return response.data;
};

export const createGenTemplate = async (workspaceId: string, data: CreateGenTemplateRequest) => {
  const response = await axios.post<GenTemplate>(
    `/api/v1/workspaces/${workspaceId}/gen-templates`,
    data,
    { withCredentials: true }
  );
  return response.data;
};

export const updateGenTemplate = async (workspaceId: string, templateId: string, data: UpdateGenTemplateRequest) => {
  const response = await axios.put<GenTemplate>(
    `/api/v1/workspaces/${workspaceId}/gen-templates/${templateId}`,
    data,
    { withCredentials: true }
  );
  return response.data;
};

export const deleteGenTemplate = async (workspaceId: string, templateId: string) => {
  const response = await axios.delete(
    `/api/v1/workspaces/${workspaceId}/gen-templates/${templateId}`,
    { withCredentials: true }
  );
  return response.data;
};

export const generateFromTemplate = async (workspaceId: string, data: GenerateFromTemplateRequest) => {
  const response = await axios.post<GenerateFromTemplateResponse>(
    `/api/v1/workspaces/${workspaceId}/gen-templates/generate`,
    data,
    { withCredentials: true }
  );
  return response.data;
};

export const getGenHistories = async (workspaceId: string, pageNum: number, pageSize: number, templateId?: string) => {
  const params = new URLSearchParams({
    pageSize: pageSize.toString(),
    pageNumber: pageNum.toString(),
  });
  if (templateId) {
    params.append('templateId', templateId);
  }
  const response = await axios.get<GenHistory[]>(
    `/api/v1/workspaces/${workspaceId}/gen-history?${params.toString()}`,
    { withCredentials: true }
  );
  return response.data;
};

export const getGenHistory = async (workspaceId: string, historyId: string) => {
  const response = await axios.get<GenHistory>(
    `/api/v1/workspaces/${workspaceId}/gen-history/${historyId}`,
    { withCredentials: true }
  );
  return response.data;
};

export const deleteGenHistory = async (workspaceId: string, historyId: string) => {
  const response = await axios.delete(
    `/api/v1/workspaces/${workspaceId}/gen-history/${historyId}`,
    { withCredentials: true }
  );
  return response.data;
};

export const listGenModels = async (workspaceId: string) => {
  const response = await axios.get<GenModel[]>(
    `/api/v1/workspaces/${workspaceId}/gen-models`,
    { withCredentials: true }
  );
  return response.data;
};

File: internal/api/handler/gen_template.go
package handler

import (
	"net/http"
	"strconv"
	"time"

	"github.com/unsealdev/unseal/internal/model"
	"github.com/unsealdev/unseal/internal/util"

	"github.com/labstack/echo/v4"
)

type CreateGenTemplateRequest struct {
	Name      string `json:"name" validate:"required"`
	Prompt    string `json:"prompt" validate:"required"`
	Provider  string `json:"provider" validate:"required"`
	Model     string `json:"model" validate:"required"`
	Modality  string `json:"modality" validate:"required"`
	ImageURLs string `json:"image_urls"`
}

type UpdateGenTemplateRequest struct {
	Name      string `json:"name" validate:"required"`
	Prompt    string `json:"prompt" validate:"required"`
	Provider  string `json:"provider" validate:"required"`
	Model     string `json:"model" validate:"required"`
	Modality  string `json:"modality" validate:"required"`
	ImageURLs string `json:"image_urls"`
}

type GetGenTemplateResponse struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Prompt      string `json:"prompt"`
	Provider    string `json:"provider"`
	Model       string `json:"model"`
	Modality    string `json:"modality"`
	ImageURLs   string `json:"image_urls"`
	CreatedAt   string `json:"created_at"`
	CreatedBy   string `json:"created_by"`
	UpdatedAt   string `json:"updated_at"`
	UpdatedBy   string `json:"updated_by"`
	WorkspaceID string `json:"workspace_id"`
}

func (h Handler) GetGenTemplates(c echo.Context) error {
	workspaceId := c.Param("workspaceId")
	pageSize := 20
	pageNumber := 1
	if ps := c.QueryParam("pageSize"); ps != "" {
		if v, err := strconv.Atoi(ps); err == nil && v > 0 {
			pageSize = v
		}
	}
	if pn := c.QueryParam("pageNumber"); pn != "" {
		if v, err := strconv.Atoi(pn); err == nil && v > 0 {
			pageNumber = v
		}
	}

	query := c.QueryParam("query")

	filter := model.GenTemplateFilter{
		WorkspaceID: workspaceId,
		PageSize:    pageSize,
		PageNumber:  pageNumber,
		Query:       query,
	}

	templates, err := h.db.FindGenTemplates(filter)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	res := make([]GetGenTemplateResponse, 0)
	for _, t := range templates {
		res = append(res, GetGenTemplateResponse{
			ID:          t.ID,
			Name:        t.Name,
			Prompt:      t.Prompt,
			Provider:    t.Provider,
			Model:       t.Model,
			Modality:    t.Modality,
			ImageURLs:   t.ImageURLs,
			CreatedAt:   t.CreatedAt,
			CreatedBy:   t.CreatedBy,
			UpdatedAt:   t.UpdatedAt,
			UpdatedBy:   t.UpdatedBy,
			WorkspaceID: t.WorkspaceID,
		})
	}

	return c.JSON(http.StatusOK, res)
}

func (h Handler) GetGenTemplate(c echo.Context) error {
	workspaceId := c.Param("workspaceId")
	if workspaceId == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "Workspace id is required")
	}

	id := c.Param("id")
	if id == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "GenTemplate id is required")
	}

	t := model.GenTemplate{WorkspaceID: workspaceId, ID: id}
	t, err := h.db.FindGenTemplate(t)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	res := GetGenTemplateResponse{
		ID:          t.ID,
		Name:        t.Name,
		Prompt:      t.Prompt,
		Provider:    t.Provider,
		Model:       t.Model,
		Modality:    t.Modality,
		ImageURLs:   t.ImageURLs,
		CreatedAt:   t.CreatedAt,
		CreatedBy:   t.CreatedBy,
		UpdatedAt:   t.UpdatedAt,
		UpdatedBy:   t.UpdatedBy,
		WorkspaceID: t.WorkspaceID,
	}

	return c.JSON(http.StatusOK, res)
}

func (h Handler) CreateGenTemplate(c echo.Context) error {
	workspaceId := c.Param("workspaceId")
	if workspaceId == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "workspace id is required")
	}
	var req CreateGenTemplateRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	if err := c.Validate(req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "Validation failed: " + err.Error(),
		})
	}

	var t model.GenTemplate
	user := c.Get("user").(model.User)

	t.WorkspaceID = workspaceId
	t.ID = util.NewId()
	t.Name = req.Name
	t.Prompt = req.Prompt
	t.Provider = req.Provider
	t.Model = req.Model
	t.Modality = req.Modality
	t.ImageURLs = req.ImageURLs
	t.CreatedAt = time.Now().UTC().String()
	t.CreatedBy = user.ID
	t.UpdatedAt = time.Now().UTC().String()
	t.UpdatedBy = user.ID

	err := h.db.CreateGenTemplate(t)

	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	return c.JSON(http.StatusCreated, t)
}

func (h Handler) DeleteGenTemplate(c echo.Context) error {
	workspaceId := c.Param("workspaceId")
	id := c.Param("id")
	if id == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "GenTemplate id is required")
	}
	template := model.GenTemplate{WorkspaceID: workspaceId, ID: id}

	existingTemplate, err := h.db.FindGenTemplate(template)

	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	user := c.Get("user").(model.User)

	if existingTemplate.CreatedBy != user.ID {
		return echo.NewHTTPError(http.StatusForbidden, "you do not have permission to delete this GenTemplate")
	}

	if err := h.db.DeleteGenTemplate(template); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusNoContent)
}

func (h Handler) UpdateGenTemplate(c echo.Context) error {
	workspaceId := c.Param("workspaceId")
	if workspaceId == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "workspace id is required")
	}
	id := c.Param("id")
	if id == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "GenTemplate id is required")
	}

	var req UpdateGenTemplateRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	if err := c.Validate(req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "Validation failed: " + err.Error(),
		})
	}

	existingTemplate, err := h.db.FindGenTemplate(model.GenTemplate{ID: id})

	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	user := c.Get("user").(model.User)

	if existingTemplate.CreatedBy != user.ID {
		return echo.NewHTTPError(http.StatusUnauthorized)
	}

	var t model.GenTemplate

	t.WorkspaceID = workspaceId
	t.ID = existingTemplate.ID
	t.Name = req.Name
	t.Prompt = req.Prompt
	t.Provider = req.Provider
	t.Model = req.Model
	t.Modality = req.Modality
	t.ImageURLs = req.ImageURLs
	t.CreatedAt = existingTemplate.CreatedAt
	t.CreatedBy = existingTemplate.CreatedBy
	t.UpdatedAt = time.Now().UTC().String()
	t.UpdatedBy = user.ID

	err = h.db.UpdateGenTemplate(t)

	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	return c.JSON(http.StatusOK, t)
}

// ListGenModels lists all available AI models from all providers with their modalities
func (h Handler) ListGenModels(c echo.Context) error {
	user := c.Get("user").(model.User)
	if user.ID == "" {
		return c.JSON(http.StatusUnauthorized, "")
	}

	// Get models from the gen service
	models, err := h.genService.ListModels()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, models)
}